### DISCUSSION TITLE
Getting MQTT data into Berry

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#19190] [Support request]
In my Berry code, I respond to MQTT requests using the function `mqtt_data()` This work just fine, when a message is sent to me, I can decode the message and take action.  

But I have noticed that every few hundred mili-second or so, if there is NO data for me, the `mqtt_data()` function is still called by Tasmota and it generates a `"Command":"Unknown"` message in the console and as a MQTT data stream message.  This generated a lot of un-needed MQTT traffic.

I'm at a loss on how to fix this... 

Below is a stripped-down version of my code, that re-creates the issue.

~~~
07:07:10.341 In mqtt_data
07:07:10.359 MQT: RSF/JUKEBOX/RESULT = {"Command":"Unknown"}
07:07:10.392 In mqtt_data
07:07:10.411 MQT: RSF/JUKEBOX/RESULT = {"Command":"Unknown"}
07:07:10.441 In mqtt_data
07:07:10.460 MQT: RSF/JUKEBOX/RESULT = {"Command":"Unknown"}
~~~

~~~

# Seeburg1.be 

    import mqtt

#- *************************************** -#
#- *************************************** -#
class SEEBURG_DRIVER : Driver

#- *************************************** -#   
    def init()
    
        tasmota.add_rule ("mqtt#connected",     /MyObj3 -> self.connected_MQTT(MyObj3)) 
        tasmota.add_rule ("mqtt#disconnected",  /MyObj4 -> self.disconnected_MQTT(MyObj4))

        print ("In init")
        mqtt.subscribe('RSF/JUKEBOX/#')  
        
    end

 #- *************************************** -#    
    def disconnected_MQTT(MyObj4)
    
        print ("MQTT Disconnect: ", MyObj4)
        #mqtt.subscribe('RSF/JUKEBOX/#') 
    end
    
#- *************************************** -#   
    def connected_MQTT(MyObj3)
    
        print("Mqtt Connected: ", MyObj3)
        #mqtt.subscribe('RSF/JUKEBOX/#') 
    end
   
#- *************************************** -# 
    def mqtt_data(topic, idx, strdata, bindata)
     
        #print ("In mqtt_data")
        
        return false
    end

end     # end of class SEEBURG_Driver : Driver

#- *************************************** -#
SEEBURG_Driver =   SEEBURG_DRIVER()
tasmota.add_driver(SEEBURG_Driver)

#- ************ The Very End ************* -#


~~~

**Comment by barbudor** [comment]
AFAIK when subscribing, it's the whole tasmota which subscribe, not a specific driver
So your mqtt_data() function is called whenever a mqtt message to any subscribed topic is received (unless someone else has claimed that message)
It up to you to analyse the topic to see if it is for you
If it is, process and return true to claim that message
If not, return false and then Berry pass the message to the next driver in queue.
If there is no more Berry driver in the queue, message is passed back to Tasmota
Tasmota check if matches any rule-event or any command,
If not, then you get the error message

**Reply by trlafleur** [reply]
Thank you... that solved the issue....

**Reply by trlafleur** [reply]
as a side question, how can I tell if MQTT is active and connected??

**Reply by barbudor** [reply]
Active as in enabled? I don't know
Connected/disconnected, I don't have a better idea for now than the rules you have already implemented
May be it would be nice to have standard call backs (mqtt_connected() /mqtt_disconnected() ) and/or a connection status (mqtt_state()) 
What do you think @s-hadinger?

**Reply by sfromis** [reply]
You have no need for those triggers for MQTT connected/disconnected at all. You can simply do the `mqtt.subscribe` without worrying about whether MQTT is connected or not (including before Wifi connected), then Berry+Tasmota will manage making the broker subscribe at the relevant points, including after outages of the connection. This is an improvement from a few months back.

Instead of that `mqtt_data` function, I'd very much recommend having the relevant function specified on the `mqtt.subscribe` providing extra flexibility like working seamlessly with multiple subscriptions, and no need to register it as a driver in Berry (unless you need it for other reasons).

This likely reduces your reasons for checking if MQTT is active and connected, but you can to some extent get clues from `tasmota.cmd('Status 6',true)`, like error for MQTT not configured, or `MqttCount` 0 for not having connected. However this does not directly tell if it can communicate.

> But I have noticed that every few hundred mili-second or so, if there is NO data for me, the mqtt_data() function is still called by Tasmota

Please give details, instead of "have noticed" without explanation. Of course, it it was only not returning true for messages you want to process....

**Reply by trlafleur** [reply]
thank you all for your great help...

**Reply by trlafleur** [reply]
Something I do not understand...
Setting return = true solves my issue... but why was Tasmota sending me an "empty" MQTT packet via `mqtt_data` in the first place??

When I subscribe to a topic, the broker is responsible for sending me a packet when it has a match to my subscription.

But in my testing, `mqtt_data` was being called every few milliseconds giving me the `RSF/JUKEBOX/RESULT = {"Command": "Unknown"}` message...

I would expect `mqtt_data` to be called ONLY when there is a valid MQTT packet from the broker to this device.

just a bit confused....

**Reply by barbudor** [reply]
Who says it is empty? 

If you print the topic and payload (strdata) you would know what exact message you are receiving
You may actually receive something

**Reply by trlafleur** [reply]
ok... all I get is the `{"Command":"Unknown"}` that Tasmota is building when I have `return = false`

still unclear to me that Tasmota is calling `mqtt_data` without a valid subscription packet. To me it looks like an error... `mqtt_data` should only be called when there is valid subscription data received....

just my two bits...

~~~
In init
In mqtt_data
topic:  RSF/JUKEBOX/RESULT
data:  {"Command":"Unknown"}
In mqtt_data
topic:  RSF/JUKEBOX/RESULT
data:  {"Command":"Unknown"}
In mqtt_data
topic:  RSF/JUKEBOX/RESULT
data:  {"Command":"Unknown"}
In mqtt_data
topic:  RSF/JUKEBOX/RESULT
data:  {"Command":"Unknown"}
~~~


~~~
# Seeburg1.be 

    import mqtt

#- *************************************** -#
#- *************************************** -#
class SEEBURG_DRIVER : Driver

#- *************************************** -#   
    def init()
    

        print ("In init")
        mqtt.subscribe('RSF/JUKEBOX/#')  
        
    end

   
#- *************************************** -# 
    def mqtt_data(topic, idx, strdata, bindata)
     
        print ("In mqtt_data")
        print ("topic: ",topic)
        print("data: ",strdata)
        tasmota.delay(1000)
        return false
    end

end     # end of class SEEBURG_Driver : Driver

#- *************************************** -#
SEEBURG_Driver =   SEEBURG_DRIVER()
tasmota.add_driver(SEEBURG_Driver)

#- ************ The Very End ************* -#
~~~

**Reply by sfromis** [reply]
I'd still suggest that you do as I already suggested. Stop all that with creating it as a driver, but specify your function as argument on the mqtt.subscribe call, preferable with another name than `mqtt_data` to avoid confusion, or in case you have other reasons to still create a driver. This means that your function is specifically coupled with the mqtt.subscribe, and nothing else. And do **always** return true if you process the message, and only false if you want to leave it to other MQTT handlers. Any message you leave may ultimately be processed as a command, which you do not seem to like :wink:

**Reply by trlafleur** [reply]
I have a new version of my code that I'm testing that work as you have suggested, but this issue still concerned me...

**Reply by s-hadinger** [reply]
I think that the answer is very simple. When you return `false`, the command is passed to Tasmota command engine (as designed), and since the command is an error, a **new** mqtt message is sent with content `{"Command":"Unknown"}`. Your `mqtt_data` is also receiving MQTT messages originating from Tasmota.

**Reply by sfromis** [reply]
And that's why I suggest to not use the driver/mqtt_data technique at all.
